<a href="https://colab.research.google.com/github/Tatiana443/Examples/blob/main/Homework_21_04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

##Условие 1: 
##Задача 1
*Скачать данные по ссылке https://www.kaggle.com/datasets/ionaskel/laptop-prices (задания в дальнейшем связаны с /kc-house-data (2).csv).
Считать данные с помощью pandas
Вывести на экран первые 5 строк*

In [ ]:
df = pd.read_csv('/laptop_price.csv', encoding='1251')
df.head()

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   
3          4   Apple  MacBook Pro  Ultrabook    15.4   
4          5   Apple  MacBook Pro  Ultrabook    13.3   

                     ScreenResolution                         Cpu   Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

                Memory                           Gpu  OpSys  Weight  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS  1.37kg   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS  1.34kg   
2            256GB SSD         Intel HD Graphics 620  No OS  1.86kg   
3            512GB SSD            AMD Radeon Pro 455  macOS  1.83kg   
4            256GB SSD  Intel Iris Plus Graphics 650  macOS  1.37kg   

   Price_euros  
0      1339.69  
1       898.94  
2       575.00  
3      2537.45  
4      1803.60

###1.2 Создать новый признак delta_renovated, который будет содержать разницу в годах между годом реновацией дома и годом постройки дома Если реновации дома не было, то в новом признаке поставьте 0.### 

In [ ]:
df = pd.read_csv('/kc-house-data (2).csv', encoding='1251')
df.head()


id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [ ]:
df.shape

(21613, 21)

In [ ]:
df['delta_renovated'] = 0
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...        1180              0   
1      7242     2.0           0     0  ...        2170            400   
2     10000     1.0           0     0  ...         770              0   
3      5000     1.0           0     0  ...        1050            910   
4      8080     1.0           0     0  ...        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  delta_renovated  
0        5650                0  
1        7639                0  
2        8062                0  
3        5000                0  
4        7503                0  

[5 rows x 22 columns]

In [ ]:
yr_delta = []

for i, row in df.iterrows():
    if row['yr_renovated']:
        yr_delta.append(row['yr_renovated']-row['yr_built'])
    else:
        yr_delta.append(0)
        
len(yr_delta)


id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...        1180              0   
1      7242     2.0           0     0  ...        2170            400   
2     10000     1.0           0     0  ...         770              0   
3      5000     1.0           0     0  ...        1050            910   
4      8080     1.0           0     0  ...        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  delta_renovated  
0        5650                0  
1        7639                0  
2        8062                0  
3        5000                0  
4        7503                0  

[5 rows x 22 columns]

In [ ]:
df['delta_renovated'] = yr_delta

df[['yr_built', 'yr_renovated', 'delta_renovated']].head()

yr_built  yr_renovated  delta_renovated
0      1955             0                0
1      1951          1991               40
2      1933             0                0
3      1965             0                0
4      1987             0                0

In [ ]:
df[df['yr_renovated'] > 0][['yr_built', 'yr_renovated', 'delta_renovated']].head()

yr_built  yr_renovated  delta_renovated
1        1951          1991               40
35       1930          2002               72
95       1946          1991               45
103      1950          2010               60
115      1900          1999               99

###1.3 Создайте признаки года продажи, месяца продажи

In [ ]:
df['yr_sale'] = df['date'].apply(lambda x: int(x[:4]))
df['mnth_sale'] = df['date'].apply(lambda x: int(x[4:6]))

df[['date', 'yr_sale', 'mnth_sale']].head()

date  yr_sale  mnth_sale
0  20141013T000000     2014         10
1  20141209T000000     2014         12
2  20150225T000000     2015          2
3  20141209T000000     2014         12
4  20150218T000000     2015          2

###1.4 Удалите признаки date, zipcode, lat, long

In [ ]:
df.drop(columns=['date', 'zipcode', 'lat', 'long'], inplace=True)
df.head()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900.0         3       1.00         1180      5650     1.0   
1  6414100192  538000.0         3       2.25         2570      7242     2.0   
2  5631500400  180000.0         2       1.00          770     10000     1.0   
3  2487200875  604000.0         4       3.00         1960      5000     1.0   
4  1954400510  510000.0         3       2.00         1680      8080     1.0   

   waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
0           0     0          3      7        1180              0      1955   
1           0     0          3      7        2170            400      1951   
2           0     0          3      6         770              0      1933   
3           0     0          5      7        1050            910      1965   
4           0     0          3      8        1680              0      1987   

   yr_renovated  sqft_living15  sqft_lot15  
0             0           1340        5650  
1          1991           1690        7639  
2             0           2720        8062  
3             0           1360        5000  
4             0           1800        7503

##Условие 2:
###2 задача
*Создайте датафрейм с клиентами:
clients = pd.DataFrame({
'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242,
2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060,
1189000207, 6821600300, 7137950720, 9510920050, 6131600255,
5428000070, 1788800910, 8100400160, 3123049142, 6306800010,
5083000375, 7920100025, 1951600150, 809001400, 339600110,
1622049154, 1099600250, 8563000110, 2768100205, 3995700435,
8861700030, 3303980210, 7731100066, 8146100580, 825069097,
3889100029, 9524100196]
})*

In [ ]:
clients = pd.DataFrame({
                        'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242,
2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
                        'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060,
1189000207, 6821600300, 7137950720, 9510920050, 6131600255,
5428000070, 1788800910, 8100400160, 3123049142, 6306800010,
5083000375, 7920100025, 1951600150, 809001400, 339600110,
1622049154, 1099600250, 8563000110, 2768100205, 3995700435,
8861700030, 3303980210, 7731100066, 8146100580, 825069097,
3889100029, 9524100196]
})
clients.head()

client_id    house_id
0       1459  8965450190
1       4684  6823100225
2       3498  5104540330
3       3942  2131701075
4       4535  1522700060

###2.1 Присоедините к таблице clients данные по домам через метод join###

In [ ]:
clients_id = clients.set_index('house_id')
df_id = df.set_index('id')

clients_id.join(df_id).reset_index().head()

index  client_id             date     price  bedrooms  bathrooms  \
0   339600110       4809  20140923T000000  395000.0         3       2.50   
1   809001400       3113  20140922T000000  925000.0         3       1.00   
2   825069097       1764  20140619T000000  770000.0         3       2.50   
3  1099600250       2242  20141202T000000  260000.0         3       1.75   
4  1189000207       2308  20141021T000000  387000.0         2       2.50   

   sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above  \
0         1610      3755     2.0           0  ...      7        1610   
1         1630      3200     1.5           0  ...      8        1630   
2         2650     40705     2.0           0  ...      9        2650   
3         1710      6400     1.0           0  ...      7        1240   
4         1170      1394     2.0           0  ...      8        1170   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1987             0    98052  47.6825 -122.097   
1              0      1912             0    98109  47.6351 -122.352   
2              0      1994             0    98053  47.6680 -122.063   
3            470      1976             0    98023  47.3036 -122.377   
4              0      2001             0    98122  47.6131 -122.297   

   sqft_living15  sqft_lot15  
0           1300        3823  
1           1710        3600  
2           2550       42625  
3           1600        6400  
4           1250        3136  

[5 rows x 22 columns]

###2.2 Присоедините к таблице clients данные по домам через метод merge
###Это нужно, чтобы понимать, какие дома покупались клиентами house_id - это индексы датафрейма с домами###

In [ ]:
merge = clients.merge(df, left_on='house_id', right_on='id')
merge.head()

client_id    house_id          id             date     price  bedrooms  \
0       1459  8965450190  8965450190  20150218T000000  295000.0         3   
1       4684  6823100225  6823100225  20150414T000000  700000.0         4   
2       3498  5104540330  5104540330  20150508T000000  679000.0         4   
3       3942  2131701075  2131701075  20141204T000000  420000.0         3   
4       4535  1522700060  1522700060  20140624T000000  518000.0         4   

   bathrooms  sqft_living  sqft_lot  floors  ...  grade  sqft_above  \
0       2.50         1500      3060     2.0  ...      7        1500   
1       1.75         1870      6000     1.0  ...      8        1670   
2       2.50         3680      7236     2.0  ...     10        3680   
3       1.75         1720      5000     1.5  ...      8        1720   
4       2.75         2520     14021     2.0  ...      9        2520   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1994             0    98006  47.5605 -122.117   
1            200      1949             0    98199  47.6435 -122.399   
2              0      2006             0    98038  47.3543 -122.005   
3              0      1932          2009    98019  47.7380 -121.983   
4              0      1999             0    98019  47.7344 -121.957   

   sqft_living15  sqft_lot15  
0           2700        7734  
1           1710        6000  
2           3310        7180  
3           1410        8300  
4           2330       14007  

[5 rows x 23 columns]

##Условие 3: 
###3 задача. Составьте несколько сводных таблиц
###3.1 Найдите среднюю стоимость домов в зависимости от количества спален. Отсортируйте от меньшей стоимости к большей

In [ ]:
df.groupby('bedrooms').agg({'price': 'mean'}).sort_values('price')

price
bedrooms              
1         3.176429e+05
2         4.013727e+05
0         4.095038e+05
3         4.662321e+05
11        5.200000e+05
4         6.354195e+05
33        6.400000e+05
5         7.865998e+05
10        8.193333e+05
6         8.255206e+05
9         8.939998e+05
7         9.511847e+05
8         1.105077e+06

###3.2 Найдите минимальную, среднюю и максимальную стоимости домов в зависимости от состояния дома

In [ ]:
df.groupby('condition').agg({'price': ['sum', 'min', 'max']}) 

price                     
                    sum       min        max
condition                                   
1          1.003295e+07   78000.0  1500000.0
2          5.629339e+07   80000.0  2555000.0
3          7.604978e+09   75000.0  7062500.0
4          2.959897e+09   89000.0  7700000.0
5          1.041723e+09  110000.0  3650000.0

###3.3 Постройте таблицу с подсчетом количества домов в данных в зависимости от вида на набережную и оценкой вида

In [ ]:
pd.crosstab(index=df['waterfront'],
            columns=df['view'])

view            0    1    2    3    4
waterfront                           
0           19489  331  955  491  184
1               0    1    8   19  135

###3.4 Каких домов в зависимости от этажности и количества спален больше?

In [37]:
pd.crosstab(index=df['floors'],
            columns=df['bedrooms'])
          

bedrooms  0    1     2     3     4    5    6   7   8   9   10  11  33
floors                                                               
1.0        4  162  1951  5455  2383  605  104   9   5   0   1   0   1
1.5        0   21   182   786   698  185   30   7   1   0   0   0   0
2.0        6   12   497  3118  3682  775  119  19   6   4   2   1   0
2.5        0    1     5    56    58   23   14   2   0   2   0   0   0
3.0        2    3   123   405    61   13    5   1   0   0   0   0   0
3.5        1    0     2     4     0    0    0   0   1   0   0   0   0

Больше одноэтажных домов с 3 спальнями (5455)

###3.5 Постройте таблицу с подсчетом медианной стоимости домов в данных в зависимости от состояния дома и оценки дома

In [46]:
pd.crosstab(index=df['condition'],
            columns=df['grade'],
            values=df['price'],
            aggfunc='median')
           
            


grade            1         3         4         5         6         7   \
condition                                                               
1          142000.0       NaN  150000.0  190000.0  255000.0  403500.0   
2               NaN  280000.0  145000.0  180000.0  235000.0  305000.0   
3               NaN   75000.0  205000.0  234475.0  265000.0  357500.0   
4               NaN       NaN  238525.0  229975.0  295000.0  390000.0   
5               NaN  262000.0       NaN  227450.0  285475.0  456000.0   

grade            8          9          10         11         12         13  
condition                                                                   
1          932500.0        NaN        NaN        NaN        NaN        NaN  
2          429000.0   715000.0  1752500.0        NaN        NaN        NaN  
3          485000.0   689000.0   890000.0  1209500.0  1807500.0  2888000.0  
4          571250.0   823500.0  1030000.0  1685000.0  2125000.0  5750000.0  
5          696000.0  1078000.0  1650000.0  2050000.0  1990000.0        NaN

In [47]:
df.pivot_table(index='condition',
               columns='grade',
               values='price',
               aggfunc='median',
               fill_value=0)

grade          1       3       4       5       6       7       8        9   \
condition                                                                    
1          142000       0  150000  190000  255000  403500  932500        0   
2               0  280000  145000  180000  235000  305000  429000   715000   
3               0   75000  205000  234475  265000  357500  485000   689000   
4               0       0  238525  229975  295000  390000  571250   823500   
5               0  262000       0  227450  285475  456000  696000  1078000   

grade           10       11       12       13  
condition                                      
1                0        0        0        0  
2          1752500        0        0        0  
3           890000  1209500  1807500  2888000  
4          1030000  1685000  2125000  5750000  
5          1650000  2050000  1990000        0